File: 프로그래머스/lv2/이선영/문자열 압축.java
import java.util.*;
class Solution {
    public int solution(String s) {
        int answer = s.length();
        String tmp = s.toString();
        
        for(int i=1; i<=s.length()/2; i++){
            StringBuilder sb = new StringBuilder();
            String base = s.substring(0,i);
            int cnt = 1;
           
            for(int j=i; j<s.length(); j+=i){
                int lastIdx = Math.min(i+j, s.length());
                String compare = s.substring(j, lastIdx);
                if(base.equals(compare)){
                    cnt++;
                }else{
                    if(cnt > 1){
                        sb.append(cnt);
                    }
                    sb.append(base);
                    base = compare;
                    cnt=1;
                }
            }
            if(cnt > 1){
                sb.append(cnt);
            }
            sb.append(base);
            answer = Math.min(answer, sb.toString().length());
            // tmp = answer == sb.toString().length() ? sb.toString() : tmp;
            // System.out.println(tmp +" " +answer );
        }
        return answer;
    }
}

File: 프로그래머스/lv2/이하은/문자열 압축.kt
import java.util.*
import kotlin.math.*

/*
압축 단위: 1..length/2 (length <= 1000)

aabbaccc (1..4)
1 a|a|b|b|a|c|c|c -> 2a2ba3c (7)
2 aa|bb|ac|cc 
3 aab|bac|cc 
4 aabb|accc

abcabcdede (1..5)
1 abcabcdede
2 ab|ca|bc|de|de -> abcabc2de (9)
3 abc|abc|ded|e -> 2abcdede (8)

1) 단위를 늘려가면서 문자열을 자른다. 
2) 연속으로 중복되는 문자열 개수를 센다. 
3) 압축 후의 문자열 길이를 구한다. 
4) 최소 길이 갱신 
*/
class Solution {
    fun solution(s: String): Int {
        var answer = s.length 
        val maxLen = s.length
        val origin = StringBuilder(s)
        val compressed = StringBuilder()
        
        for(unit in 1..maxLen/2){
            var prev = origin.substring(0, unit)
            var cnt = 1
            
            for(i in unit until maxLen step unit){
                val now = origin.substring(i, min(i + unit, maxLen))
                
                // 연속으로 중복되는 문자열 개수 카운트 
                if(prev == now){
                    cnt++
                }else{
                    if(cnt >= 2){
                        compressed.append("$cnt$prev")
                    }else{
                        compressed.append(prev)
                    }
                    
                    // 그 다음 케이스를 위한 변수 갱신
                    prev = origin.substring(i, min(i + unit, maxLen))
                    cnt = 1
                }
            }
            
            // 남은 문자열 처리 
            if(cnt >= 2){
                compressed.append("$cnt$prev")
            }else{
                compressed.append(prev)
            }

            // 최소 길이 갱신 
            answer = min(answer, compressed.length)
            compressed.setLength(0)
        }
        
        return answer 
    }
}